core/src/hash/hasher.rs

```diff
@@ -0,0 +1,360 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::{
+    hash::{
+        BuildHasherDefault,
+        Hasher,
+    },
+    marker::PhantomData,
+};
+use ink_prelude::vec::Vec;
+
+mod markers {
+    /// Prevents users from implementing certain traits or generic types.
+    pub trait Sealed {}
+
+    pub enum Sha2x256Marker {}
+    pub enum Keccakx256Marker {}
+    pub enum Blake2x256Marker {}
+    pub enum Blake2x128Marker {}
+    pub enum TwoxMarker {}
+
+    impl Sealed for Sha2x256Marker {}
+    impl Sealed for Keccakx256Marker {}
+    impl Sealed for Blake2x256Marker {}
+    impl Sealed for Blake2x128Marker {}
+    impl Sealed for TwoxMarker {}
+}
+
+/// Generic cryptographic hasher.
+pub struct CryptoHasher<T, A>
+where
+    T: markers::Sealed,
+{
+    /// The accumulator buffer.
+    ///
+    /// A bytes buffer that is used to accumulate a hashing state.
+    accumulator: A,
+    /// Tricks compiler into thinking that `Self` uses `T`.
```

Comment:
seems over-documenting, it is not even a public member

Reply:
I can understand your argument here and I am undecided about this. So far I always documented minor things like this also to teach new Rust people along reading source code. Less questions, more progress for new Rustaceans here and there.